lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1146 @@
+) -> Result<(), PeerHandleError>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    // Init channels
+    let (reader_cmd_tx, reader_cmd_rx, writer_cmd_tx, writer_cmd_rx) = init_channels();
+
+    // Generate a new ID that represents this connection
+    let conn_id = ID_COUNTER.fetch_add(1, Ordering::AcqRel);
+    let socket_addr = stream.peer_addr().unwrap();
+
+    // Init TcpReader, TcpWriter, TcpDisconnectooor
+    let writer_stream = stream.try_clone().unwrap();
+    let disconnector_stream = stream.try_clone().unwrap();
+    let tcp_reader = TcpReader(stream);
+    let tcp_writer = TcpWriter(writer_stream);
+    let tcp_disconnector = TcpDisconnectooor(disconnector_stream);
+
+    // Init SyncSocketDescriptor
+    let mut descriptor = SyncSocketDescriptor::new(
+        conn_id,
+        tcp_disconnector,
+        reader_cmd_tx.clone(),
+        writer_cmd_tx.clone(),
+    );
+
+    // Init Reader and Writer
+    let mut reader: Reader<CMH, RMH, L, UMH> = Reader::new(
+        tcp_reader,
+        peer_manager.clone(),
+        descriptor.clone(),
+        reader_cmd_rx,
+        writer_cmd_tx,
+    );
+    let mut writer: Writer<CMH, RMH, L, UMH> = Writer::new(
+        tcp_writer,
+        peer_manager.clone(),
+        descriptor.clone(),
+        reader_cmd_tx,
+        writer_cmd_rx,
+    );
+
+    // Notify the PeerManager of the new connection depending on its ConnectionType.
+    //
+    // - If Ok, spawn the Reader and Writer threads.
+    // - If Ok and Outbound, additionally queue up the initial data.
+    // - If Err, disconnect the TcpStream and do not spawn the worker threads.
+    //
+    // In all cases, return the result of the call into the PeerManager.
+    match conn_type {
+        ConnectionType::Inbound => {
+            peer_manager.new_inbound_connection(descriptor.clone(), Some(to_net(socket_addr)))
+        }
+        ConnectionType::Outbound(their_node_id) => peer_manager
+            .new_outbound_connection(their_node_id, descriptor.clone(), Some(to_net(socket_addr)))
+            .map(|initial_data| {
+                // PeerManager accepted the outbound connection; queue up the
+                // initial WriteData WriterCommand.
+                let bytes_pushed = descriptor.send_data(&initial_data, true);
+                // This should always succeed since send_data just pushes data
+                // into the writer_cmd channel and the writer_cmd channel always
+                // starts out completely empty. If pushing the initial 10s of
+                // bytes into the writer_cmd channel fails, something is very
+                // wrong; probably a programmer error.
+                if bytes_pushed != initial_data.len() {
+                    panic!("The initial write should always succeed");
+                }
```

Comment:
Even in that case, there would be space in the `writer_cmd` channel, so `send_data()` would succeed since all it does is push the `WriteData` request into the channel. There isn't a case where this initial `send_data()` call doesn't succeed

Reply:
ahh i see, the reader hasn't been spawned yet. 👌